### ISSUE TITLE
New lint: `ref` of reference

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#6644]
### What it does

Checks for a `ref` binding which creates a reference to a reference.

### Categories (optional)

- Kind: style

Reduces de-referencing, less code without `ref`.

### Drawbacks

None.

### Example

```rust
if let Some(ref x) = Some(&1_i32) {
    dbg!(x); // x is a &&i32
}
```

Could be written as:

```rust
if let Some(x) = Some(&1_i32) {
    dbg!(x);
}
```

Name could be `ref_of_ref` or `redundant_ref`

**Comment by pag4k** [comment]
Hey! I would like to give a shot at this issue. I don't have much experience with this kind of contribution so it might take a while. :-)

**Comment by pag4k** [comment]
Hello! :-) I got this working for very simple cases like the provided example, but I have a few questions:
1. I assumed this has to be done on LatePass, do you think did is correct?
2. To what extent do we want to handle complex cases? I struggle to find a general way to handle any cases.
3. Do we want to also handle `ref mut`?
Thank you!

**Comment by camsteffen** [comment]
You will need a late pass for type checking. It should be easy to detect all cases. You just implement `check_pat`, find any `ref` binding and look at the type using `cx.typeck_results()`.

Unfortunately I think we need to take an extra measure to avoid false positives ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=16143f3e41dfc782370d9d54af3260b0)), which might make this not a good first issue. We might avoid false positives by simply avoiding `ref mut`, but it would be better to use `ExprUseVisitor` to ensure that all usages of the binding start with a deref.

About the lint name, I think it should include the word "binding" to make it clear that it does not lint code like `&&x`. So I'd call it `ref_binding_of_ref`.

**Comment by pag4k** [comment]
Thank you for the reply. It is an interesting problem, but I think I'll try to find another one to experience to whole process with an easy one first. :-)

**Comment by camsteffen** [comment]
This is part of the `needless_borrow` lint.
